src/JsonApi/Serializer/CollectionNormalizer.php

```diff
@@ -0,0 +1,207 @@
+                ),
+            ],
+        ];
+
+        if ($paginated) {
+            $returnDataArray['links']['first'] = IriHelper::createIri(
+                $parsed['parts'],
+                $parsed['parameters'],
+                $this->pageParameterName,
+                1.
+            );
+
+            $returnDataArray['links']['last'] = IriHelper::createIri(
+                $parsed['parts'],
+                $parsed['parameters'],
+                $this->pageParameterName,
+                $lastPage
+            );
+
+            if (1. !== $currentPage) {
+                $returnDataArray['links']['prev'] = IriHelper::createIri(
+                    $parsed['parts'],
+                    $parsed['parameters'],
+                    $this->pageParameterName,
+                    $currentPage - 1.
+                );
+            }
+
+            if ($currentPage !== $lastPage) {
+                $returnDataArray['links']['next'] = IriHelper::createIri(
+                    $parsed['parts'],
+                    $parsed['parameters'],
+                    $this->pageParameterName,
+                    $currentPage + 1.
+                );
+            }
+        }
+
+        $identifier = null;
+        foreach ($data as $item) {
+            $normalizedItem = $this->normalizer->normalize($item, $format, $context);
+
+            if (!isset($normalizedItem['data'])) {
+                throw new RuntimeException(
+                    'data key expected but not found during JSON API normalization'
+                );
+            }
+
+            $normalizedItem = $normalizedItem['data'];
+
+            $relationships = [];
+
+            if (isset($normalizedItem['relationships'])) {
+                $relationships = $normalizedItem['relationships'];
+
+                unset($normalizedItem['relationships']);
+            }
+
+            foreach ($normalizedItem['attributes'] as $property => $value) {
+                $propertyMetadata = $this->propertyMetadataFactory->create($resourceClass, $property);
+
+                if ($propertyMetadata->isIdentifier()) {
+                    $identifier = $normalizedItem['attributes'][$property];
+                }
+            }
+
+            $items = [
+                'type' => $resourceMetadata->getShortName(),
+                // The id attribute must be a string
+                // http://jsonapi.org/format/#document-resource-object-identification
+                'id' => (string) $identifier ?? '',
+                'attributes' => $normalizedItem['attributes'],
+            ];
+
+            if ($relationships) {
+                $items['relationships'] = $relationships;
+            }
+
+            $returnDataArray['data'][] = $items;
+        }
+
+        if (is_array($data) || $data instanceof \Countable) {
+            $returnDataArray['meta']['totalItems'] = $data instanceof PaginatorInterface ?
+                (int) $data->getTotalItems() :
```

Comment:
It's not necessary (JSON doesn't make the distinction between int and float, it only has a number type).